### DISCUSSION TITLE
List and function of ManufacturerSpecificCluster in ZHA

### DISCUSSION THREAD

**Discussion by MattWestb** [zigpy/zigpy#823] [Ideas]
ManufacturerSpecificCluster is used in Zigbee then one manufacture is implanting no Zigbee standard functions that can being used in one Zigbee system.

IKEA is using one and sending commands then adding scenes in there lights to making the scenes ignoring the light level and only changing the color of the light.

Some known ManufacturerSpecificCluster is: 

tuya: 0xef00 (is used for sending DPs), 0xe000 and 0xe001 (locks being used for changing device propitiates) both is **not** in zigbee range for ZMC !!

IKEA: 0xFC7C, 0xFC57.

Philips: 0xfc00, 0xfc01, 0xfc02 and 0xfc03 Looks like last firmware is using the last one and the other is outdated).

I think some function can in the future being implanted in the system that  is using some of theis without quirk is they is being percent on the device for adding more standard functions from common used devices.

Knowing more one ? Pleas adding them to the list then ZHA dont have one.

**Comment by MattWestb** [comment]
tuya is using 0xEF00 (61184) for there DP commands to TS0601 devices (tunneling MQTT commands to there cloud) and some hybrid devices that using normal Zigbee cluster for commands and DP for advanced config.

### "Normal" tuya Zigbee devices:
#### Manufactur cluster:

- 0xE000 Not known (used in some plugs and new TS004X remotes)
- 0xE001 Switch mode cluster: ✅
  - 0xD010 Power On Behavior (t.enum8) 
    - 0x00 = Off 
    - 0x01 = On 
    - 0x02 = Last state 
  - 0xD011 Unknown and is used in some switches then casting "tuya magic"  
  - 0xD030 Switch mode (t.enum8) ✅
    - 0x00 Toggle ✅
    - 0x01 State ✅
    - 0x02 Momentary ✅

- 0xE002 is being used in some sensors with attribute: ☑️
  - 0xD00A: ("alarm_temperature_max", t.uint16_t), ☑️
  - 0xD00B: ("alarm_temperature_min", t.uint16_t), ☑️
  - 0xD00C: ("alarm_humidity_max", t.uint16_t), ☑️
  - 0xD00E: ("alarm_humidity_min", t.uint16_t), ☑️
  - 0xD00F: ("alarm_humidity", ValueAlarm), ☑️
    - 0x00 Min alarm ☑️
    - 0x01 Max alarm ☑️
    - 0x02 Amarm off ☑️
  - 0xD006: ("temperature_humidity", ValueAlarm), ☑️
    - 0x00 Min alarm ☑️
    - 0x01 Max alarm ☑️
    - 0x02 Amarm off ☑️
  - 0xD010: {"unknown", t.uint8_t}, ☑️
- 0xFD02 Not know (used in some switch modules)
- 0xEF00 tuya DP cluster `External Switch-Type` ⬜ / ✅ (Some hybrid devices is using it for advance Switch configuring)
   - 0x04 Send DP command
   - 0x05 Reporting DP commands

Known DP commands for  `External Switch-Type`: 
- 0 Unknown
- 1 Toggle
-  2 Momentary
- 3 Rotary
- 4 Auto Confi

#### Extra "custom" commands and  attribute added to ZCL cluster:
Commands:
- 0x0006 OnOff Cluster ✅
  - 0xFC Rotate commands ✅
    - 0x00 Rotate right ✅
    - 0x01 Rotate  left ✅
    - 0x02 Rotate  stop ✅
  - 0xFD "Scen commands" ✅
    - 0x00 Single press ✅
    - 0x01 Double press ✅
    - 0x02 Long press ✅
 
- 0x0008 Level Control Cluster ⬜
  - 0x00F0 tuya Level Command, level: (t.uint16_t) 0-1000, transtime: (t.uint16_t) ⬜

Attributes:
- 0x0000 Basic Cluster
  - 0xfffe Unknown but looks being one part of the "tuya magic" for some devics. ⬜

- 0x0006 OnOff Cluster ✅
  - 0x5000 Back Light Switch (t.enum8)
  - 0x8000 Child_Lock (t.Bool) 0/1 ✅
  - 0x8001 Back Light Mode (t.enum8) ✅
    - 0x00 BL Mode 0 ✅
    - 0x01 BL Mode 1 ✅
    - 0x02 BL Mode 2 ✅
  - 0x8002 Power On State (t.enum8) ✅
    - 0x00 = Off ✅
    - 0x01 = On ✅
    - 0x02 = Last state ✅
  - 0x8003 Over Current Alarm (t.Bool) 0/1 (Default 0x00)
    - 0x00 = Over Current OK
    - 0x01 = Over Current Alarm
  - 0x8004 Switch Operation Mode (t.enum8) ✅
    - 0x00 = Command Mode (Light OnOff / Dimmer commands) ✅
    - 0x01 = Event Mode (tuya "scene" commands) ✅

- 0x0008 Level Control Cluster ⬜
  - 0xF000 "tuya Level control" (t.uint16_t) 0-1000 ⬜
  - 0xfc00 Min and Max dimming Level (t.uint16_t) (default 0x01ff) ⬜
    - 0xAABB
AA = The first 8 Bit= Min dim level 0x03XX = 1%, 0xfcXX = 99% and 0xffXX = Not Set
BB = The last 8 Bit = Assumed being the same but for Max dim level.
  - 0xfc02 Bulb type	(uint8) 0x00–0xFF (default 0x00)
    - 0x00 LED
    - 0x01 Incandescent
    - 0x02 Halogen
  - 0xfc03 Scr state (uint8) 0x00–0xFF (default 0x01)
  - 0xfc04 Current percentage	(uint8)  0x00–0xFF (default 0x01)
  - 0xfc05 Min percentage (uint8) 0x00–0xFF	(default 0x01)

Looks like the dim min and max can having different attribute (combined 0xfc00 or spate 0xFC03/0xFC04) on different devices.
- 0x0102 Window covering cluster: ⬜
  - 0xF000 Window cover status (t.enum8) ⬜
    - 0x00 Up/open ⬜
    - 0x01 Down/close ⬜
    - 0x02 Stop ⬜
  - 0xF001 Calibration (t.enum8) ⬜
  - 0xF002 Motor reversed (t.enum8) ⬜
  - 0xF003 Calibration time (t.uint16_t) ⬜

- 0x0502 IasWd Cluter
  - 0xF000 (int8)
    - 0 Disarm the alarm
    - 1 Sound alarm
    - 2 Light alarm
    - 3 Sound and light alarm

✅ = Implanted in tuya INIT.
☑️ = Used in some device quirk and is candidate for moving into INIT.
⬜ = Used in local quirk by users and is candidate for moving into INIT.

**Comment by MattWestb** [comment]
@miit86 i have found some functions of your first _tuya PITA cluster_ `0xE001 Switch mode cluster` but i dont knowing if its attributes or commands sent to the switch but i think is attributes.

Like trying defining it and look if your device is responding on  it ?

**Reply by miit86** [reply]
@MattWestb 

0xE001 Switch mode cluster:
  0xD030 (t.enum8)
    0x00 Toggle
    0x01 State
    0x02 Momentary

How to you want me to define them? Im always available for testing Tuya Pita devices

**Reply by MattWestb** [reply]
Try do it like the power on state you was doing but with one new cluster.
I dont knowing how to defining one new cluster only updating attributes like you was doing but look on the `ts130f.py` that is adding extra cluster on one "normal" Zigbee device.

PS: I have adding back light to tuya OnOff cluster (and renamed it) and its looks working but my power strip is not having any back light but its reporting the attribute OK. https://github.com/zigpy/zha-device-handlers/pull/1105.
Also mode switch for some switches that also is working.

**Reply by MattWestb** [reply]
https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/tuya/ts0201_neo.py is defining attributes and also making  one new custom cluster !!

**Comment by MattWestb** [comment]
Zigbee is having one 0x0007 clster _OnOff configuration_ that is very rear used but it good.
The spec you can finding in the Zigbee Cluster Library (ZCL) [ZCL Ver 7](https://github.com/Koenkk/zigbee-herdsman/blob/master/docs/Zigbee%20Cluster%20Library%20Specification%20v7.pdf) 3.9 (download the PDF the its too large for view in GitHub).

Its supporting monetary (2 stable state) toggle (2 states 1 stable) and multi function (OnOff Dimmer).
And also inverting command or only sending toggle for both states (1 and 2).

I think 2 of the tuya attributes i compatible but the last not ;-((

**Comment by MattWestb** [comment]
## Some interesting findings in new updated tuya ZCL R8 docks
### Level control cluster
level control cluster(0x0008)	
Level Control Max Min(0xFC00)	
Level Control Bulb Type(0xFC02)	
Level Control Scr State(0xFC03)	
Level Control Current Percent(0xFC04)	
Level Control Min Percent(0xFC05)

Private cluster
Attribute:
| ID |	Name | Data type | Range | Defualt value |
|--|--|--|--|--|
|0xFC00|	Level control max min|	uint16 -0x21|	0x0000 - 0xffff	|0x01ff|
|0xFC02|	Level control bulb type|	uint8 -0x20|	0x00–0xFF|	0x00|
|0xFC03|	Level control scr state|	uint8 -0x20|	0x00–0xFF|	0x01|
|0xFC04|	Level control current percentage|	uint8 -0x20|	0x00–0xFF|	0x01|
|0xFC05|	Level control min percentage|	uint8 -0x20|	0x00–0xFF|	0x01|

Special instructions:
The Zigbee smart dimmer switch (no neutral wire required) uses a thyristor to realize switching and dimming actions. Use the above private attributes to store and communicate related information.

0xFC00 is used to store the current maximum and minimum brightness values.
0xFC02 is used to distinguish the bulb type.
0xFC03 is used to store the status of the thyristor.
0xFC04 is used to store the current dimming percentage.
0xFC05 is used to store the minimum brightness percentage.
The minimum brightness value is used to adapt to different types of LED lights with different wattages, and the thyristor status is used for dual control functions.

### On/Off cluster

Attributes
| ID |	Name | Data type | Range | Defualt value |
|--|--|--|--|--|
|0x8003|	Over Current Alarm|	Bool-0x10|	0/1|	0x00|

Over-current alerts
(1) Use an external metering chip to measure the real-time current. If the real-time current read exceeds the set maximum operating current, an overcurrent alarm will be triggered, and the Over Current Alarm (private attribute) of On Off Cluster is set to 1.

(2) Turn off the relay that is being turned on, and immediately report the overcurrent alarm status Over Current Alarm to the gateway.

### IAS WD Cluster

0xF000	Private attribute: a custom private attribute used to display prompts on the app.	int8 -0x20 true/false	
0: Disarm the alarm.
1: Sound alarm.
2: Light alarm.
3: Sound and light alarm.

Special instructions:
1. Alarm, currently only recognizes the Waring Mode field of the gateway's Start Waring.
Stop means stop, not Stop. For example, Fire will alarm and execute according to the parameters stored locally.

2. After receiving the alarm and stop alarm commands, the alarm will send the zone status change notification command, with the lowest bit Alarm0 bit set to 0 or 1, and then report to report the zone status attribute.

**Reply by MattWestb** [reply]
@miit86 Over current alarm attribute found on OnOff cluster but not tested.
- 0x8003 Over Current Alarm (Bool) 0/1 (Default 0x00)
   - 0x00 = Over Current OK
   - 0x01 = Over Current Alarm

**Comment by javicalle** [comment]
I have a couple of this Tuya switches (Moes branding):
* [ZigBee Smart Wall Light Switch](https://www.moeshouse.com/collections/tuya-zigbee-switch/products/zigbee-smart-wall-light-switch-with-neutral-wire-or-no-neutral-wire-wiring-no-capacitor-needed-smart-life-tuya-2-3-way-muilti-control-association-works-with-alexa-google-home-hub-required-gray)

They (as others tuya switches) are using a DP for controlling the `power_on_state` and `backlight_mode`.
In my tests the `backlight_mode` accepts the following values:
* off = 0x00
* light_when_on = 0x01
* light_when_off = 0x02
* freeze = 0x03

The freeze value sets the backlight to the current state and only can be tested with a double switch device.

The strange thing is that on the product website it is announced that the `backlight` can be configured as on/off, but I have not found a value that allows it to be defined as on, it is only possible if you apply the freeze value while the backlight is on.
![Moes backlight control](https://cdn.shopify.com/s/files/1/0095/4079/6497/products/zigbee-smart-wall-light-switch-with-neutral-wire-or-no-neutral-wire-wiring-no-capacitor-needed-smart-lifetuya-23-way-muilti-control-association-works-with-alexa-742123_600x600.jpg?v=1621053966)

I don't know if those settings are specific to Moes devices or other Tuya devices are using them.

Grettings.

**Reply by MattWestb** [reply]
Thanks for information and feedback that is very interesting for our users and devs !!

Moes is only one "German brand" or somthing like that that is buying all kinds of tuya products and branding them. They have doing it with nearly all TRVs but behind is always one manufacture with one name.
With TRVs i have structurating them with the manufacturer names but the old Moes is still left in the quirks then renaming it is braking the test and i cant fixing it.

I my thinking all tuya devices is belonging in the tuya quirk folder and not being splitted as Z2M have doing (ZHA is having one LIDL but its only for fixing some lights).

Back too attribute / commands !!!
Some tuya Touch (DP) switches is having backlight mode that can being on / off or dimmed and changing color then being on or off.
Very likely they is using same (DP) commands for doing different things or have implanted new ones that we is not knowing.

If you is finding some Zigbee Cluster Library (ZCL) compliment commands like power on default on / off / last i think its best way is putting them on the same attribute as the ZCL ones then ZHA devs is starting implanting GUI elements for ZCL functions and then its possible getting them in GUI without making any changes in the quirk and in ZHA.
And attribute that is not fitting i think they shall being tagged with tuya manufacturer so ZHA is knowing its no standard ones and also its possible making the same in ZHA GUI with the tagged attribute in the future.

tuya scene switches is also having some special for reporting how many gangs they is having on startup with DP but its not the normal DP that  is used for other devices but you can finding them in tuya dev  decks then you is needing them.

One thing that is little discussed in the background is moving tuya DP things one level higher (ZHA or one "tuya DP module" in HA) for getting better implementation in HA that is one great PITA with TRVs and only need adding one quirks that is piping the DP to the module and all is done in zigbee layer and the upper is doing the rest. The same can being used for ZHA, Tasmota Z2M and more if they is liking that and being much free implanting the GUI element in HA. I dont knowing if its being done but i think its the best way getting good functions on tuya DP devices in (Z)HA.

Great thanks for your work done with switches and dimmers and hope you is getting more great code in our systems that making all working better in the end !!

**Reply by javicalle** [reply]
I have to say that I have mixed feelings with everything related to Tuya devices.

On the one hand I can understand and share the need to make the devices work in HA (or any other ecosystem).

But on the other hand, I feel like Tuya is doing a great business selling non-standard 'things' as Zigbee devices. And that we are doing the dirty work saying "it does not matter, we will make it work as standard".

That causes people to keep buying and complaining to HA, Zigpy, whoever that his new devices are not working, diverting the pressure from the manufacturer to the developers/maintainers.

I don't know, maybe we should give a different message and say "if you want it to work, don't buy Tuya devices, buy from manufacturers who care about following the Zigbee standard"

**Reply by MattWestb** [reply]
It was the same with IKEA was not working so well with other Zigbee system in the beginning but thy was fixing most of there issues and being one of the heavy board members of Zigbee Alliance.
Philips HUE is also problematic then they is not reporting status from there lights and need being pulled that is not standard and is not aging neighbours and children that have leaving the network.
Then Xiaomi is making all in there way and braking all.

tuya "normal Zigbee" is not so bad then they is mostly using Silabs SDK that is working OK and getting bug fixed (but IKEA is not rolling the latest so still having bad bugs in there devices like battery draining in remotes).

tuya is being board member in Zigbee Alliance and that is very heavy in the market but they is only certifying device if the customer is demanding it like LIDL have done.
Over 90% is not certified and that's bad for open systems.

The DP device is little different then the same tuya MCU can using WiFi, BT, 433 or Zigbee and only need changing the radio module and its working out of the box. tuya is having 2 hardware for Zigbee DP devices and 2 software one for sleeping end device and one for router devices so its pretty basic.
I think only the communicating module in the MCU is need getting other config for working with one Zigbee module but i dont  knowing if the manufacturer string is in the Zigbee module or if its coming from the MCU and the Zigbee module is "clean".

tuya is having some very nice touch dimmer but then its DP devices its not possible making binding i Zigbee so they is not working if the host system is not on line = one no go for my production system.

Ony "Zigbee" side its good and bad then tuya is implanting som things no standard way like power on state with different attribute that is not needed and can being dome with the standard ones maped OK and adding some with extended manufacturer attributes and all was 110% better. The same with power metering and switch configuration as you can see in the first posting.

One bad with tuya is Zigbee updates /certificat is not working well. LIDL was seling one 3 plug power strip with 4 USB and was certificated and new ones have getting one updated firmware that is not certificated and its only having one end point and switching all 3 plugs so waiting for some user is sniffing it for see what magic they have doing (hidden end points or using unknown commands).

And i agree tuya is so large so they is the number one in IOT for us then Xiaomi is behind and no one can comparing with the gigant. Hopefully they is doing better products in the future that is working with our system without making  tons of extra code and revers engineering there devices.

The problem is getting cheap devices for our systems is not easy to finding like light switches, buch jäger its great devices but is costing 10 times one tuya switch so its out of range for most user.

IKEA is cheap and normally working good but is not having wall switches only remotes.
Xiaomi is doing there "click click" and only the Opple serie is working (if implanting it in ZHA) as light switches.
So in the end most user is going for tuya devices.

tuya TRVs is very problematic in ZHA then they is having so much functions that  is not possible squeezing thru ZCL commands and also on the scheduling side 3 or more different implementations so its no tuya standard that  can helping on the road.

Z2M is having great time then they is only throwing tuya commands thru MQTT and dont need the ZCL for getting the functions in HA. deCONZ have completely freeing all PR then reworking there device base for getting user adding there device with file editing in GUI (no tuya DP device for 1/2 year = ver bad for the system).

At least we have getting most DP blind controller, TRVs and switches up and running but i have guíving up the TRVs then no devs can helping getting the working quirks merged then we cant writing the tests for it.

But one the good side tuya is not making the life boring and always finding new interesting things with them !!!

**Comment by MattWestb** [comment]
I have found the Child Lock on the OnOff cluster in Z2M but i cant finding witch data type its using.
```
fromZigbee.js

    },
    ts011f_plug_child_mode: {
        cluster: 'genOnOff',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const property = (0x8000).toString(); // 32768
            if (msg.data.hasOwnProperty(property)) {
                const value = msg.data[property];
                return {child_lock: value ? 'LOCK' : 'UNLOCK'};
            }
        },
    },



toZigbee.js

    ts011f_plug_child_mode: {
        key: ['child_lock'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('genOnOff', {0x8000: {value: value === 'LOCK', type: 0x10}});
        },
    },
```
@javicalle can you see if its one t.enum8 or it somthing else ?

**Reply by javicalle** [reply]
This code `value: value === 'LOCK'` is assigning a boolean as value.
This other code `child_lock: value ? 'LOCK' : 'UNLOCK'` is converting a boolean (`value`) to the `LOCK` or `UNLOCK` values.

From what I understand, in Zigbee the value is managed as boolean. I'm going to investigate this `type: 0x10` part that seems to define the type of the value.

**Reply by javicalle** [reply]
Looking at usage in the rest of the library, my guess is that `type: 0x10` refers to a boolean (`t.Bool`).
To be sure, I'm trying to find the implementation of the Cluster class at the library.

**Reply by MattWestb** [reply]
I also thinking its one bool.
Perhaps i can getting one user trying it :-))

Its tuya Zigbee device so "standard" is not CSA with this devices.

Thanks for looking on it !!!

**Reply by javicalle** [reply]
Without looking at the rest of the code, my guess would be that the `child_lock` is handled by writing a boolean to the 0x8000 attribute
However, in the `zha-device-handlers` library, other values of this type (`MOESBHT_CHILD_LOCK_ATTR`, `SITERWELL_CHILD_LOCK_ATTR`, `ZONNSMART_CHILD_LOCK_ATTR`, ...) seem to be handling it with a `t.uint8_t` even though it seems clear that the possible values are 0/1

**Reply by MattWestb** [reply]
The OnOff attribute 0x8000 is for "normal" tuya Zigbee device and as you knowing tuya very "flexible" then its coming to coding there things. The MCU / DP for TRV is one other large mess and is all you can do without Zigbee standard.

The user have start making code for testing it and we will knowing very son i think. 👍

**Reply by Adminiuga** [reply]
ZigBee type 0x10 check the specs.
In tozigbee.ts you can see the type

**Reply by javicalle** [reply]
I was looking for somewhere where value types were declared, similar to what is in `zha-device-handlers`:
https://github.com/zigpy/zha-device-handlers/blob/039a31f1ca15821e800bdb347eb47fdb02455f31/zhaquirks/tuya/__init__.py#L193-L201

I haven't found where this posible types are declarated in `zigbee-herdsman-converters`. Maybe I have to look at zigpy,

**Reply by Adminiuga** [reply]
Check zcl specification, it defines the types and corresponding type between type_id and python class is defined in zigpy/zcl/foundation.py

**Reply by javicalle** [reply]
Caught:
```python
        0x10: ("Boolean", t.Bool, Discrete),
```

Thanks for guiding me!

**Comment by MattWestb** [comment]
Have found the command used for hybrid devices with  0xEF00 tuya DP cluster:

- 0xEF00 tuya DP cluster `External Switch-Type` ⬜ / ✅ (Some hybrid devices is using it for advance Switch configuring)
   - 0x04 Send DP command
   - 0x05 Active DP Status Report

Known DP commands for  `External Switch-Type`: 
- 0 Unknown
- 1 Toggle
-  2 Momentary
- 3 Rotary
- 4 Auto Confi

Can being of interest for @javicalle and our HB friend @kkossev

Edit: Looks one more command is being found but i think i have wrong with the DPs then its more ones and is grouped that  i need finding out.

- 0x06 Active DP Status Report "Alt"

Link to Z2M PR https://github.com/Koenkk/zigbee-herdsman/pull/519

I have seen somthing in tuya dev docks but i must finding it agen then its lost in the storm ;-((

**Reply by MattWestb** [reply]
Need feeding  the DPs being used but was finding little more interesting  things in https://developer.tuya.com/en/docs/iot/tuya-zigbee-module-uart-communication-protocol?id=K9ear5khsqoty.

Its possible subscribing reporting from the MCU for one or more DP !!!

If have interest  look little more of the commands they is using so getting little more knowledge if there protocol

**Comment by MattWestb** [comment]
One Norwegian brand have released some interesting information how they is faking tuya ZBGW and have released one PDF with our tuya magic sequence https://github.com/dresden-elektronik/deconz-rest-plugin/issues/6198#issuecomment-1193817322.

Also is tuya calling the unknown attribute `attributeReportingStatus (0xFFFE)`.

If some TRVs is needing it i thinking it is needed for more or less all tuya MCU device but it need being tested on many before assign it "globally" but i think its making more good in the future if we start adding it for all new tuya MCU and "normal Zigbee" devices.

Holla @javicalle  some interesting reading for you in the PDF attached from the above issue [Connecte Smart Thermostat.pdf](https://github.com/zigpy/zigpy/files/9559071/Connecte.Smart.Thermostat.pdf).

By the way is Valencia not liking my and have bad weather then im visiting on my Holiday ;-(

**Reply by javicalle** [reply]
I am also convinced that most of the hacks could be applied by default to all Tuya devices (the spell cast, the `NO_MANUFACTURER_ID`,...) 
I also suspect that they will not harm the 'old' devices and that the new ones will surely do need so.

In the PDF are collected most of our adventures. Looks like someone else has been through the same thing (they mention integration with [Homey](https://homey.app/))

Offtopic: I'm sorry for your bad experience (I'm sure the bad weather didn't help). Valencia has some very nice places to visit (the Old City, the "Ciudad de las ciencias"...), but like anywhere your experience will depend on where you were 🤷🏻‍♂️

**Reply by MattWestb** [reply]
Then im back for Gandia i like trying updating the TS004F Quirk so its using the spell for all device classes then we is having 2 devices plus one "general" and one upcoming that is working (need the spell) then i have 2 of the device and the user for the new one is good tester but the only problem is how to do it general for more classes without duplicating the code all the time. For MCU devices is only implanting it on the tuya cluster but most normal devices dont have that and one dev was trying using the basic cluster but it was not working then paring device and need it very early in the paring.

Off topic: IKEA VLC was having the Styrbar in stainless steel for good price and 2 large **BLACK** (not the coding check  tool !!!) Coffey cups so only 30€ for 2 of each and the holiday is great also then its raining :-))). Shall taking one more look on the old city in VLC then have doing it one time but its very nice and i like what i have seen !!

**Reply by javicalle** [reply]
> one dev was trying using the basic cluster but it was not working then paring device and need it very early in the paring

The `EnchantedDevice` implementation is casting the spell right after the device initialization:
```python
    def __init__(self, *args, **kwargs):
        """Initialize with task."""
        super().__init__(*args, **kwargs)
        self._init_device_task = asyncio.create_task(self.spell())
```

The spell its just reading from the cluster (the `Basic` cluster here):
```python
    async def spell(self) -> None:
        """Initialize device so that all endpoints become available."""
        attr_to_read = [4, 0, 1, 5, 7, 0xFFFE]
        basic_cluster = self.endpoints[1].in_clusters[0]
        await basic_cluster.read_attributes(attr_to_read)
```

Maybe, we can put that call just in the `Basic` cluster initialization. 
That way it would not wait for the rest of the clusters/endpoints on the device to initialize.
I don't know if at that point the code would be functional, but I think it's worth a try.

My proposal, a new `EnchantedBasicCluster` class that extends the `Basic` cluster (`from zigpy.zcl.clusters.general import Basic` will be needed):
```python
class EnchantedBasicCluster(Basic):
    """Class for enchanted Tuya devices which needs to be unlocked by casting a 'spell'."""

    def __init__(self, endpoint, is_server: bool = True):
        """Initialize with task."""
        super().__init__(endpoint, is_server)
        self._init_device_task = asyncio.create_task(self.spell())

    async def spell(self) -> None:
        """Initialize device so that all endpoints become available."""
        attr_to_read = [4, 0, 1, 5, 7, 0xFFFE]
        await self.read_attributes(attr_to_read)
```

Not sure if will work because tha call is inside the init method, so cluster would not be initialized yet to read the attributes.

Now in the rogue device, define the quirk replacement with the new Basic cluster, like that:
```python
    signature = {
        ENDPOINTS: {
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.SMART_PLUG,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Time.cluster_id,
                    .../...
                ],
                OUTPUT_CLUSTERS: [Ota.cluster_id],
            }
        },
    }

    replacement = {
        ENDPOINTS: {
            1: {
                DEVICE_TYPE: zha.DeviceType.ON_OFF_LIGHT,
                INPUT_CLUSTERS: [
                    EnchantedBasicCluster,
                    Time.cluster_id,
                    .../...
                ],
                OUTPUT_CLUSTERS: [Ota.cluster_id],
            }
        }
    }
```

Pingme when you are going to test the code and I will try to assist you.

**Comment by wang70880** [comment]
Hi Matt, I would like to know how you infer the payload of these commands in these manufacture-specific clusters. For example, you can know that the 0x0006 (OnOFF cluster) has a command 0xFC (Rotate) which payload 0x00/0x01/0x02. I am curious how you know the payload.

**Reply by MattWestb** [reply]
Some have using tuya dev portal for looking and is getting all used command (for tuya MCU is the best).
For TS004F devices some users have using debug and getting the payload.
For the rotate i was using wireshark and sniffing the ROK and then testing and comparing with debug log.
Attributes is little more tricky if not have some documented of them and then its best using tuya ZBGW and sniffing for getting function and payload and then can testing in ZHA. Also scanning the devices with ZHA-Toolkit is working well with tuya (normal Zigbee devices not MCU) for getting attribute and format of the payload but some devices is being crazy after doing that and need being deleted and new added for working OK.

Also looking in tuya teck dock its possible finding some but is not always true then tuya is little flexible.
If Z2M have digging is normally the best and easiest way getting the relevant information and implanting the same (they is doing the same with or findings so its very OK).

**Reply by wang70880** [reply]
Sorry but I am confused about some parts.

1.  Does the tuya dev portal mean the Tuya cloud development platform? Do you mean to refer to the development documentation?
2. What is "ROK"?

**Reply by MattWestb** [reply]
1A: Yes having one tuya ZBGW and connecting it tuya dev platform.
1B: Tuy dev dock you finding on internet its not writing all but some good information can being found.
2: Its our code name for tuya Smart Rotating Knob (TS004F).

**Reply by wang70880** [reply]
Thanks a lot. I will investigate further. So for the identification of manufacturer-specific commands, currently zigpy mainly focuses on Tuya products right? For other brands like Xiaomi, SmartThings, Philips, does the same method apply for command identification?

**Reply by MattWestb** [reply]
If one device is Zigbee 3 and ZCL R8 compatible it shall working out of the box with all functions like most  IKEA devices on latest firmware is doing.
tuya "normal" Zigbee have manny not standard commands / attributes and tuya MCU devise is only using Zigbee as one comport sending commands to the cloud.
Xiaomi is very un Zigbee 3 and normally not working at all without quirks.
The right way is using the standard ZCL R8 and if need other functions use one manufacture cluster like IKEA have doing with the Symfonis 2 then its using Matter ZCL that is not in Zigbee but is cumming in next update of it then IKEA can changing the cluster number to standard and all is working OK.

So all no Zigbee standard commands / attributes shall being on one manufacture cluster tagged with the manufacture number = all other manufacture devices is ignoring it.

By the way of bad devices you was missing Osram and you have getting all that having problem with standard and functionality ;-)

**Reply by Hedda** [reply]
Not sure if they will help, but FYI, there's more links in the README.md from zha-device-handlers repo (including Tuya-specific ones):

https://github.com/zigpy/zha-device-handlers/blob/dev/README.md

https://github.com/zigpy/zha-device-handlers/blob/dev/README.md#what-are-these-specifications

https://github.com/Koenkk/zigbee-herdsman/blob/master/docs/Zigbee%20Connection%20Standard_Tuya%20Smart_Documentation.pdf

https://www.zigbee2mqtt.io/advanced/support-new-devices/02_support_new_tuya_devices.html

https://developer.tuya.com/en/docs/iot/tuya-zigbee-module-uart-communication-protocol?id=K9ear5khsqoty

**Reply by wang70880** [reply]
> Some have using tuya dev portal for looking and is getting all used command (for tuya MCU is the best). For TS004F devices some users have using debug and getting the payload. For the rotate i was using wireshark and sniffing the ROK and then testing and comparing with debug log. Attributes is little more tricky if not have some documented of them and then its best using tuya ZBGW and sniffing for getting function and payload and then can testing in ZHA. Also scanning the devices with ZHA-Toolkit is working well with tuya (normal Zigbee devices not MCU) for getting attribute and format of the payload but some devices is being crazy after doing that and need being deleted and new added for working OK.
> 
> Also looking in tuya teck dock its possible finding some but is not always true then tuya is little flexible. If Z2M have digging is normally the best and easiest way getting the relevant information and implanting the same (they is doing the same with or findings so its very OK).

Hi Matt, sorry to bother you again. I reviewed your sentences, and I think you mentioned two methods for the command payload identification.

1. Use Tuya Dev portal.
2. Monitor devices by sniffing and collecting the Zigbee traffic logs.

I am still confused about one question: You mentioned "scanning the devices with ZHA-Toolkit", what is the ZHA-Toolkit? Any link for it?

**Reply by wang70880** [reply]
> https://developer.tuya.com/en/docs/iot/tuya-zigbee-module-uart-communication-protocol?id=K9ear5khsqoty

Thanks for your info. I will check the referred Tuya documentation.

**Reply by MattWestb** [reply]
Its  utility that can making many extra things that is not implanted in ZHA and one is scanning all / selected endpoints of one device for attributes its supporting )only working on good Zigbee 3 devices) and you is getting the list of them in the HA log.

Look on the devs git.
https://github.com/mdeweerd/zha-toolkit


Number 2 can being done with wire shark but then its very low level data but can being  great.
Also setting debug for ZHA and Zigpy you can getting all data the device is sending but you cant finding all things then not knowing if tuya is making some special setup from the device then its needed one sniff from pairing with one tuya ZVGW and analyzing it and implaniting it in ZHA / zigpy.

**Reply by Hedda** [reply]
> Look on the devs git.
> https://github.com/mdeweerd/zha-toolkit

Also check out related discussion about zha-toolkit here:

https://github.com/zigpy/zigpy/discussions/901

+ see the zha-toolkit thread in the Home Assistant community forum:

https://community.home-assistant.io/t/zha-toolkit-a-big-set-of-zigbee-commands-on-top-of-zha-zigpy/373346

**Comment by coen1111** [comment]
I have two types of Tuya devices which are in wall dimmers. Support for the dimmer functionality is added in [3190](https://github.com/zigpy/zha-device-handlers/pull/3190). However, the Dataopoints in the Tuya manufacturing cluster are not added. Which does not enable changing the External Switch-Type as required by my application. An example implementation is present in Zigbee2MQTT [tuya.ts](https://github.com/RafaelDuncan/zigbee-herdsman-converters/blob/3d43ed3cee2606fd4ab517e215ecf9059b4cc7d8/src/devices/tuya.ts#L1614). I want to extend this support to ZHA. I tried making a custom quirks but did not succeed. Using the zha_toolkit I now try to create the command manually, however I did not succeed yet.

This is my action:

```
action: zha_toolkit.zcl_cmd
data:
  ieee: a4:c1:38:dc:2b:c5:a5:61  # Replace with your device's IEEE address
  endpoint_id: 1                 # The endpoint ID of the device, usually 1 for most devices
  cluster: 61184                 # 0xEF00 in decimal
  cmd: 4                         # The command ID for Send DP
  args:
    - 4                          # DP ID for External Switch-Type (check device specifics)
    - 0                          # Data type (0 for raw/unknown, may vary)
    - 2                          # Data value (set to 2 for Momentary)
```

What am I overlooking?
